pydis_core/exts/source.py

```diff
@@ -0,0 +1,198 @@
+        else:
+            identifier = _TagIdentifierStub.from_string(argument.lower())
+            if identifier in tags_cog.tags:
+                return identifier, _SourceType.tag
+
+        escaped_arg = escape_markdown(argument)
+
+        raise commands.BadArgument(
+            f"Unable to convert '{escaped_arg}' to valid command{', tag,' if show_tag else ''} or Cog."
+        )
+
+    def _get_source_link(self, source_item: object, source_type: _SourceType) -> tuple[str, str, int | None]:
+        """
+        Build GitHub link of source item, return this link, file location and first line number.
+
+        Raise BadArgument if `source_item` is a dynamically-created object (e.g. via internal eval).
+        """
+        if source_type == _SourceType.text_command or source_type == _SourceType.core_command:
+            source_item = inspect.unwrap(source_item.callback)
+            src = source_item.__code__
+            filename = src.co_filename
+        elif source_type == _SourceType.tag:
+            tags_cog = self.bot.get_cog("Tags")
+            filename = tags_cog.tags[source_item].file_path
+        else:
+            src = type(source_item)
+            try:
+                filename = inspect.getsourcefile(src)
+            except TypeError:
+                raise commands.BadArgument("Cannot get source for a dynamically-created object.")
+
+        if source_type != _SourceType.tag:
+            try:
+                lines, first_line_no = inspect.getsourcelines(src)
+            except OSError:
+                raise commands.BadArgument("Cannot get source for a dynamically-created object.")
+
+            lines_extension = f"#L{first_line_no}-L{first_line_no+len(lines)-1}"
+        else:
+            first_line_no = None
+            lines_extension = ""
+
+        # Handle tag file location differently than others to avoid errors in some cases
+        if not first_line_no:
+            file_location = Path(filename)
+        elif source_type == _SourceType.core_command:
+            package_location = metadata.distribution("pydis_core").locate_file("") / "pydis_core"
+            internal_location = Path(filename).relative_to(package_location).as_posix()
+            file_location = "pydis_core/" + internal_location
+        elif source_type == _SourceType.core_cog:
+            package_location = metadata.distribution("pydis_core").locate_file("") / "pydis_core" / "exts"
+            internal_location = Path(filename).relative_to(package_location).as_posix()
+            file_location = "pydis_core/exts/" + internal_location
+        else:
+            file_location = Path(filename).relative_to(Path.cwd()).as_posix()
+
+        repo = self.github_repo if source_type != _SourceType.core_command else BOT_CORE_REPO
+
+        if source_type == _SourceType.core_command or source_type == _SourceType.core_cog:
+            version = f"v{metadata.version('pydis_core')}"
+        else:
+            version = "main"
```

Comment:
Could this perhaps use bot.sha or bot.version in order to pin to the same commit as the currently running code? This will prevent link decay.

Reply:
cfc254d03e57f8ec6ae050b5a7704a2cb8b4f5c0